File: api/src/main/java/org/openmrs/PersonAttribute.java
/**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */
package org.openmrs;

import java.io.Serializable;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Comparator;
import java.util.Date;

import org.apache.commons.lang3.StringUtils;
import org.hibernate.search.annotations.Analyzer;
import org.hibernate.search.annotations.Boost;
import org.hibernate.search.annotations.DocumentId;
import org.hibernate.search.annotations.Field;
import org.hibernate.search.annotations.Fields;
import org.hibernate.search.annotations.Indexed;
import org.hibernate.search.annotations.IndexedEmbedded;
import org.openmrs.api.context.Context;
import org.openmrs.api.db.hibernate.search.LuceneAnalyzers;
import org.openmrs.util.OpenmrsClassLoader;
import org.openmrs.util.OpenmrsUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * A PersonAttribute is meant as way for implementations to add arbitrary information about a
 * user/patient to their database. PersonAttributes are essentially just key-value pairs. However,
 * the PersonAttributeType can be defined in such a way that the value portion of this
 * PersonAttribute is a foreign key to another database table (like to the location table, or
 * concept table). This gives a PersonAttribute the ability to link to any other part of the
 * database A Person can have zero to n PersonAttribute(s).
 * 
 * @see org.openmrs.PersonAttributeType
 * @see org.openmrs.Attributable
 */
@Indexed
public class PersonAttribute extends BaseChangeableOpenmrsData implements java.io.Serializable, Comparable<PersonAttribute> {
	
	public static final long serialVersionUID = 11231211232111L;
	
	private static final Logger log = LoggerFactory.getLogger(PersonAttribute.class);
	
	// Fields
	@DocumentId
	private Integer personAttributeId;

	@IndexedEmbedded(includeEmbeddedObjectId = true)
	private Person person;

	@IndexedEmbedded
	private PersonAttributeType attributeType;

	@Fields({
			@Field(name = "valuePhrase", analyzer = @Analyzer(definition = LuceneAnalyzers.PHRASE_ANALYZER), boost = @Boost(8f)),
			@Field(name = "valueExact", analyzer = @Analyzer(definition = LuceneAnalyzers.EXACT_ANALYZER), boost = @Boost(4f)),
			@Field(name = "valueStart", analyzer = @Analyzer(definition = LuceneAnalyzers.START_ANALYZER), boost = @Boost(2f)),
			@Field(name = "valueAnywhere", analyzer = @Analyzer(definition = LuceneAnalyzers.ANYWHERE_ANALYZER))
	})
	private String value;
	
	/** default constructor */
	public PersonAttribute() {
	}
	
	public PersonAttribute(Integer personAttributeId) {
		this.personAttributeId = personAttributeId;
	}
	
	/**
	 * Constructor for creating a basic attribute
	 * 
	 * @param type PersonAttributeType
	 * @param value String
	 */
	public PersonAttribute(PersonAttributeType type, String value) {
		this.attributeType = type;
		this.value = value;
	}
	
	/**
	 * Shallow copy of this PersonAttribute. Does NOT copy personAttributeId
	 * 
	 * @return a shallows copy of <code>this</code>
	 */
	public PersonAttribute copy() {
		return copyHelper(new PersonAttribute());
	}
	
	/**
	 * The purpose of this method is to allow subclasses of PersonAttribute to delegate a portion of
	 * their copy() method back to the superclass, in case the base class implementation changes.
	 * 
	 * @param target a PersonAttribute that will have the state of <code>this</code> copied into it
	 * @return Returns the PersonAttribute that was passed in, with state copied into it
	 */
	protected PersonAttribute copyHelper(PersonAttribute target) {
		target.setPerson(getPerson());
		target.setAttributeType(getAttributeType());
		target.setValue(getValue());
		target.setCreator(getCreator());
		target.setDateCreated(getDateCreated());
		target.setChangedBy(getChangedBy());
		target.setDateChanged(getDateChanged());
		target.setVoidedBy(getVoidedBy());
		target.setVoided(getVoided());
		target.setDateVoided(getDateVoided());
		target.setVoidReason(getVoidReason());
		return target;
	}
	
	/**
	 * Compares this PersonAttribute object to the given otherAttribute. This method differs from
	 * {@link #equals(Object)} in that this method compares the inner fields of each attribute for
	 * equality. Note: Null/empty fields on <code>otherAttribute</code> /will not/ cause a false
	 * value to be returned
	 * 
	 * @param otherAttribute PersonAttribute with which to compare
	 * @return boolean true/false whether or not they are the same attributes
	 * <strong>Should</strong> return true if attributeType value and void status are the same
	 */
	@SuppressWarnings("unchecked")
	public boolean equalsContent(PersonAttribute otherAttribute) {
		boolean returnValue = true;
		
		// these are the methods to compare.
		String[] methods = { "getAttributeType", "getValue", "getVoided" };
		
		Class attributeClass = this.getClass();
		
		// loop over all of the selected methods and compare this and other
		for (String methodAttribute : methods) {
			try {
				Method method = attributeClass.getMethod(methodAttribute);
				
				Object thisValue = method.invoke(this);
				Object otherValue = method.invoke(otherAttribute);
				
				if (otherValue != null) {
					returnValue &= otherValue.equals(thisValue);
				}
				
			}
			catch (NoSuchMethodException e) {
				log.warn("No such method for comparison " + methodAttribute, e);
			}
			catch (IllegalAccessException | InvocationTargetException e) {
				log.error("Error while comparing attributes", e);
			}

		}
		
		return returnValue;
	}
	
	// property accessors
	
	/**
	 * @return Returns the person.
	 */
	public Person getPerson() {
		return person;
	}
	
	/**
	 * @param person The person to set.
	 */
	public void setPerson(Person person) {
		this.person = person;
	}
	
	/**
	 * @return the attributeType
	 */
	public PersonAttributeType getAttributeType() {
		return attributeType;
	}
	
	/**
	 * @param attributeType the attributeType to set
	 */
	public void setAttributeType(PersonAttributeType attributeType) {
		this.attributeType = attributeType;
	}
	
	/**
	 * @return the value
	 */
	public String getValue() {
		return value;
	}
	
	/**
	 * @param value the value to set
	 */
	public void setValue(String value) {
		this.value = value;
	}
	
	/**
	 * @see java.lang.Object#toString()
	 * <strong>Should</strong> return toString of hydrated value
	 */
	@Override
	public String toString() {
		Object o = getHydratedObject();
		if (o instanceof Attributable) {
			return ((Attributable) o).getDisplayString();
		} else if (o != null) {
			return o.toString();
		}
		
		return this.value;
	}
	
	/**
	 * @return the personAttributeId
	 */
	public Integer getPersonAttributeId() {
		return personAttributeId;
	}
	
	/**
	 * @param personAttributeId the personAttributeId to set
	 */
	public void setPersonAttributeId(Integer personAttributeId) {
		this.personAttributeId = personAttributeId;
	}
	
	/**
	 * Will try to create an object of class 'PersonAttributeType.format'. If that implements
	 * <code>Attributable</code>, hydrate(value) is called. Defaults to just returning getValue()
	 * 
	 * @return hydrated object or getValue()
	 * <strong>Should</strong> load class in format property
	 * <strong>Should</strong> still load class in format property if not Attributable
	 */
	@SuppressWarnings("unchecked")
	public Object getHydratedObject() {
		
		if (getValue() == null) {
			return null;
		}
		
		try {
			Class c = OpenmrsClassLoader.getInstance().loadClass(getAttributeType().getFormat());
			try {
				Object o = c.newInstance();
				if (o instanceof Attributable) {
					Attributable attr = (Attributable) o;
					return attr.hydrate(getValue());
				}
			}
			catch (InstantiationException e) {
				// try to hydrate the object with the String constructor
				log.trace("Unable to call no-arg constructor for class: " + c.getName());
				return c.getConstructor(String.class).newInstance(getValue());
			}
		}
		catch (Exception e) {
			
			// No need to warn if the input was blank
			if (StringUtils.isBlank(getValue())) {
				return null;
			}
			
			log.warn("Unable to hydrate value: " + getValue() + " for type: " + getAttributeType(), e);
		}
		
		log.debug("Returning value: '" + getValue() + "'");
		return getValue();
	}
	
	/**
	 * Convenience method for voiding this attribute
	 * 
	 * @param reason
	 * <strong>Should</strong> set voided bit to true
	 */
	public void voidAttribute(String reason) {
		setVoided(true);
		setVoidedBy(Context.getAuthenticatedUser());
		setVoidReason(reason);
		setDateVoided(new Date());
	}
	
	/**
	 * @see java.lang.Comparable#compareTo(java.lang.Object)
	 * <strong>Should</strong> return negative if other attribute is voided
	 * <strong>Should</strong> return negative if other attribute has earlier date created
	 * <strong>Should</strong> return negative if this attribute has lower attribute type than argument
	 * <strong>Should</strong> return negative if other attribute has lower value
	 * <strong>Should</strong> return negative if this attribute has lower attribute id than argument
	 * <strong>Should</strong> not throw exception if attribute type is null
	 * Note: this comparator imposes orderings that are inconsistent with equals
	 */
	@Override
	public int compareTo(PersonAttribute other) {
		DefaultComparator paDComparator = new DefaultComparator();
		return paDComparator.compare(this, other);
	}
	
	/**
	 * @since 1.5
	 * @see org.openmrs.OpenmrsObject#getId()
	 */
	@Override
	public Integer getId() {
		
		return getPersonAttributeId();
	}
	
	/**
	 * @since 1.5
	 * @see org.openmrs.OpenmrsObject#setId(java.lang.Integer)
	 */
	@Override
	public void setId(Integer id) {
		setPersonAttributeId(id);
		
	}
	
	/**
	 Provides a default comparator.
	 @since 1.12
	 **/
	public static class DefaultComparator implements Comparator<PersonAttribute>, Serializable {

		private static final long serialVersionUID = 1L;
		
		@Override
		public int compare(PersonAttribute pa1, PersonAttribute pa2) {
			int retValue;
			if ((retValue = OpenmrsUtil.compareWithNullAsGreatest(pa1.getAttributeType(), pa2.getAttributeType())) != 0) {
				return retValue;
			}
			
			if ((retValue = pa1.getVoided().compareTo(pa2.getVoided())) != 0) {
				return retValue;
			}
			
			if ((retValue = OpenmrsUtil.compareWithNullAsLatest(pa1.getDateCreated(), pa2.getDateCreated())) != 0) {
				return retValue;
			}
			
			if ((retValue = OpenmrsUtil.compareWithNullAsGreatest(pa1.getValue(), pa2.getValue())) != 0) {
				return retValue;
			}
			
			return OpenmrsUtil.compareWithNullAsGreatest(pa1.getPersonAttributeId(), pa2.getPersonAttributeId());
		}
	}
	
}

File: api/src/main/kotlin/org/openmrs/PersonAttribute.kt
/**
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
 * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
 *
 * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
 * graphic logo is a trademark of OpenMRS Inc.
 */
package org.openmrs

import jakarta.persistence.Cacheable
import org.apache.commons.lang3.StringUtils
import org.codehaus.jackson.annotate.JsonIgnore
import org.hibernate.annotations.Cache
import org.hibernate.annotations.CacheConcurrencyStrategy
import org.hibernate.envers.Audited
import org.hibernate.search.mapper.pojo.automaticindexing.ReindexOnUpdate
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.AssociationInverseSide
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.DocumentId
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.FullTextField
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.Indexed
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.IndexedEmbedded
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.IndexingDependency
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.ObjectPath
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.PropertyValue
import org.openmrs.api.context.Context
import org.openmrs.api.db.hibernate.search.SearchAnalysis
import org.openmrs.util.OpenmrsClassLoader
import org.openmrs.util.OpenmrsUtil
import org.slf4j.LoggerFactory
import java.io.Serializable
import java.util.*

/**
 * A PersonAttribute is meant as way for implementations to add arbitrary information about a
 * user/patient to their database. PersonAttributes are essentially just key-value pairs. However,
 * the PersonAttributeType can be defined in such a way that the value portion of this
 * PersonAttribute is a foreign key to another database table (like to the location table, or
 * concept table). This gives a PersonAttribute the ability to link to any other part of the
 * database A Person can have zero to n PersonAttribute(s).
 *
 * @see org.openmrs.PersonAttributeType
 * @see org.openmrs.Attributable
 */
@Indexed
@Audited
@Cacheable
@Cache(usage = CacheConcurrencyStrategy.READ_WRITE)
class PersonAttribute() : BaseChangeableOpenmrsData(), Serializable, Comparable<PersonAttribute> {

    @DocumentId
    var personAttributeId: Int? = null

    @IndexedEmbedded(includeEmbeddedObjectId = true)
    @AssociationInverseSide(inversePath = [ObjectPath([PropertyValue(propertyName = "attributes")])])
    var person: Person? = null

    @IndexedEmbedded
    @IndexingDependency(reindexOnUpdate = ReindexOnUpdate.SHALLOW)
    var attributeType: PersonAttributeType? = null

    @FullTextField(name = "valuePhrase", analyzer = SearchAnalysis.PHRASE_ANALYZER)
    @FullTextField(name = "valueExact", analyzer = SearchAnalysis.EXACT_ANALYZER)
    @FullTextField(name = "valueStart", analyzer = SearchAnalysis.START_ANALYZER, searchAnalyzer = SearchAnalysis.EXACT_ANALYZER)
    @FullTextField(name = "valueAnywhere", analyzer = SearchAnalysis.ANYWHERE_ANALYZER, searchAnalyzer = SearchAnalysis.EXACT_ANALYZER)
    var value: String? = null

    constructor(personAttributeId: Int?) : this() {
        this.personAttributeId = personAttributeId
    }

    /**
     * Constructor for creating a basic attribute
     *
     * @param type PersonAttributeType
     * @param value String
     */
    constructor(type: PersonAttributeType?, value: String?) : this() {
        this.attributeType = type
        this.value = value
    }

    /**
     * Shallow copy of this PersonAttribute. Does NOT copy personAttributeId
     *
     * @return a shallows copy of this
     */
    fun copy(): PersonAttribute = copyHelper(PersonAttribute())

    /**
     * The purpose of this method is to allow subclasses of PersonAttribute to delegate a portion of
     * their copy() method back to the superclass, in case the base class implementation changes.
     *
     * @param target a PersonAttribute that will have the state of this copied into it
     * @return Returns the PersonAttribute that was passed in, with state copied into it
     */
    protected fun copyHelper(target: PersonAttribute): PersonAttribute {
        target.person = person
        target.attributeType = attributeType
        target.value = value
        target.creator = creator
        target.dateCreated = dateCreated
        target.changedBy = changedBy
        target.dateChanged = dateChanged
        target.voidedBy = voidedBy
        target.voided = voided
        target.dateVoided = dateVoided
        target.voidReason = voidReason
        return target
    }

    /**
     * Compares this PersonAttribute object to the given otherAttribute. This method differs from
     * [equals] in that this method compares the inner fields of each attribute for
     * equality. Note: Null/empty fields on `otherAttribute` /will not/ cause a false value to be returned
     *
     * @param otherAttribute PersonAttribute with which to compare
     * @return boolean true/false whether or not they are the same attributes
     * Should return true if attributeType value and void status are the same
     */
    fun equalsContent(otherAttribute: PersonAttribute): Boolean {
        var returnValue = true

        // these are the methods to compare.
        val methods = arrayOf("getAttributeType", "getValue", "getVoided")

        val attributeClass = this.javaClass

        // loop over all of the selected methods and compare this and other
        for (methodAttribute in methods) {
            try {
                val method = attributeClass.getMethod(methodAttribute)

                val thisValue = method.invoke(this)
                val otherValue = method.invoke(otherAttribute)

                if (otherValue != null) {
                    returnValue = returnValue && otherValue == thisValue
                }

            } catch (e: NoSuchMethodException) {
                log.warn("No such method for comparison $methodAttribute", e)
            } catch (e: Exception) {
                log.error("Error while comparing attributes", e)
            }
        }

        return returnValue
    }

    /**
     * Will try to create an object of class 'PersonAttributeType.format'. If that implements
     * Attributable, hydrate(value) is called. Defaults to just returning getValue()
     *
     * @return hydrated object or getValue()
     * Should load class in format property
     * Should still load class in format property if not Attributable
     */
    fun getHydratedObject(): Any? {
        if (value == null) {
            return null
        }

        try {
            val c = OpenmrsClassLoader.getInstance().loadClass(attributeType?.format)
            try {
                val o = c.getDeclaredConstructor().newInstance()
                if (o is Attributable<*>) {
                    return o.hydrate(value)
                }
            } catch (e: InstantiationException) {
                // try to hydrate the object with the String constructor
                log.trace("Unable to call no-arg constructor for class: ${c.name}")
                return c.getConstructor(String::class.java).newInstance(value)
            }
        } catch (e: Exception) {
            // No need to warn if the input was blank
            if (StringUtils.isBlank(value)) {
                return null
            }

            log.warn("Unable to hydrate value: $value for type: $attributeType", e)
        }

        log.debug("Returning value: '$value'")
        return value
    }

    /**
     * Convenience method for voiding this attribute
     *
     * @param reason
     * Should set voided bit to true
     */
    fun voidAttribute(reason: String?) {
        voided = true
        voidedBy = Context.getAuthenticatedUser()
        voidReason = reason
        dateVoided = Date()
    }

    override fun toString(): String {
        val o = getHydratedObject()
        return when (o) {
            is Attributable<*> -> o.displayString
            null -> value ?: ""
            else -> o.toString()
        }
    }

    override fun getId(): Int? = personAttributeId

    override fun setId(id: Int?) {
        personAttributeId = id
    }

    /**
     * @see java.lang.Comparable#compareTo(java.lang.Object)
     * Should return negative if other attribute is voided
     * Should return negative if other attribute has earlier date created
     * Should return negative if this attribute has lower attribute type than argument
     * Should return negative if other attribute has lower value
     * Should return negative if this attribute has lower attribute id than argument
     * Should not throw exception if attribute type is null
     * Note: this comparator imposes orderings that are inconsistent with equals
     */
    override fun compareTo(other: PersonAttribute): Int {
        return DefaultComparator().compare(this, other)
    }

    /**
     * Provides a default comparator.
     * @since 1.12
     */
    class DefaultComparator : Comparator<PersonAttribute>, Serializable {
        override fun compare(pa1: PersonAttribute, pa2: PersonAttribute): Int {
            var retValue = OpenmrsUtil.compareWithNullAsGreatest(pa1.attributeType, pa2.attributeType)
            if (retValue != 0) {
                return retValue
            }

            retValue = pa1.voided!!.compareTo(pa2.voided!!)
            if (retValue != 0) {
                return retValue
            }

            retValue = OpenmrsUtil.compareWithNullAsLatest(pa1.dateCreated, pa2.dateCreated)
            if (retValue != 0) {
                return retValue
            }

            retValue = OpenmrsUtil.compareWithNullAsGreatest(pa1.value, pa2.value)
            if (retValue != 0) {
                return retValue
            }

            return OpenmrsUtil.compareWithNullAsGreatest(pa1.personAttributeId, pa2.personAttributeId)
        }

        companion object {
            private const val serialVersionUID = 1L
        }
    }

    companion object {
        private const val serialVersionUID = 11231211232111L
        private val log = LoggerFactory.getLogger(PersonAttribute::class.java)
    }
}
